core/trino-main/src/main/java/io/trino/dispatcher/FailedDispatchQueryFactory.java

```diff
@@ -58,7 +57,6 @@ public FailedDispatchQuery createFailedDispatchQuery(Session session, String que
         BasicQueryInfo queryInfo = failedDispatchQuery.getBasicQueryInfo();
 
         queryMonitor.queryCreatedEvent(queryInfo);
-        queryMonitor.queryImmediateFailureEvent(queryInfo, toFailure(throwable));
```

Comment:
Why having it here was causing race condition? `queryCreatedEvent` is called just before anyway.

Why query has to be registered via `queryTracker.addQuery`?

Reply:
I have provided relevant information in the extended commit message, like you asked in the other comment. Let me know whether it's clear now. If not, I will try again to describe it better.

Reply:
Consider adding `QueryTracker` as an arg to `createFailedDispatchQuery`. This way you don't have to add `queryMonitor` to `DispatchManager`